Source/DafnyCore/Backends/Dafny/DafnyCodeGenerator.cs

```diff
@@ -1264,7 +1263,36 @@ protected override ILvalue SeqSelectLvalue(SeqSelectExpr ll, ConcreteSyntaxTree
     }
 
     protected override ILvalue MultiSelectLvalue(MultiSelectExpr ll, ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new UnsupportedFeatureException(ll.Origin, Feature.MultiDimensionalArrays);
+      if (ll.Indices.Count > 16) {
+        throw new UnsupportedFeatureException(ll.Origin, Feature.ArraysWithMoreThan16Dims);
+      }
+
+      if (!ll.Array.Type.IsArrayType) {
+        throw new InvalidOperationException();
+      }
+
+      EmitExpr(
+        ll.Array, false,
+        EmitCoercionIfNecessary(
```

Comment:
I would prefer to refactor this large sub-expression into a variable, .e.g `var w = `